### ISSUE TITLE
Support of rendering to another repository

### ISSUE THREAD

**Issue by dan1el-k** [akuity/kargo#1354]
# Checklist

* [x] I've searched the issue queue to verify this is not a duplicate feature request.
* [x] I've pasted the output of `kargo version`, if applicable.
* [x] I've pasted logs, if applicable.

# Proposed Feature

Rendering to a another repository. (Service Repo concept)

# Motivation

As of version `0.3.0` kargo, just supports rendering in branches of the same repository as the source `helm` or `kustomize` files exists.

This is fine and supported for:
- Single application repos: containing manifests of one application
- Mono application repos: containing manifests of multiple application within the same repository. Those could be rendered in to a Stage branch in multiple subfolder

What I would like to have is the support of:
- Service repos: repo for the Business service, containing all the kargo deployment objects for this project, and which is used as `Rendered Target` and furthermore as deployment source for ArgoCD. 
To bundle multiple application rendered manifests in one repo, however having the original manifests still in separate repos.

With that, this `Service repo`could be a pure pipeline/infrastructure repo with no manual changes need after setup, and the `application repos`(multiple, one for each microservice) are used as a source and working repo for the team/engineers.

# Suggested Implementation

Support of following new parameters

- `readRepoURL`: currently `repoURL`, mandatory
- `readBranch`: optional as of now
- `writeBranch`: optional as of now
- `writeRepoURL`: **optional, new**

Full example
```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: dev
  namespace: kargo-demo-service
spec:
  subscriptions:
    warehouse: demo-kustomize
  promotionMechanisms:
    gitRepoUpdates:
    - readRepoURL: https://gh-enterprise.com/my-org/kargo-demo-kustomize
      readBranch: main
      writeRepoURL: https://gh-enterprise.com/my-org/kargo-demo-service
      writeBranch: stage/dev
      render:
        images:
        - image: artifacts.my-comp/docker/my-org/cortex-template-kustomize-application/nginx
```

**Comment by dan1el-k** [comment]
In the enterprise company I am working in, teams uses different repo patterns, based on size and complexity of the applications.

- `Mono repo pattern`: Application source code, Kubernetes manifests, value overlays, project docu and description, pipeline settings - all in one repo. Mainly used for smaller aka 1 microservice only business services
- `Mono manifest repo pattern`: Application source code in separate repos per microservice, Kubernetes manifests value overlays, project docu and description, pipeline settings - all in a single manifest repo. Either used for mid-sized services to save the effort for separate manifest repos per microservice and where the team is small, or also used for service with very several similiar microservices using the same manifest with different values per microservice and stage.
- `Service/Project repo pattern`: Application source code in one repo with a sibling kubernetes manifest repo with value overlays for each microservice. Project docu and description, pipeline settings in a service/project repo. Mainly used for larger and/or more complex application with 3+ microservices in there application stack.

![kargo_monorepo](https://github.com/akuity/kargo/assets/34611072/39110ae8-b4b5-4343-b179-28741a4cd0d0)

![kargo_mono-manifest](https://github.com/akuity/kargo/assets/34611072/e2c4ff21-0382-454b-94ae-d3bc04b30f90)

![kargo_servicerepo](https://github.com/akuity/kargo/assets/34611072/fee517ba-5107-4d3d-873f-730b253bd140)

---
While the first 2 patterns are perfectly supported by kargo, I am specifically targeting the 3rd one with this request.

Yes, this could potentially be achieved as well using Kustomize remote base or an Helm umbrella chart with dependencies or via Git Submodules out of the service repo, as I have requested with [my other ticket](https://github.com/akuity/kargo-render/issues/232) in "kargo-render" project. 
I do believe however, that a native support to render to a remote repo is easier to configure and grasp for the enduser

**Comment by MarkusNeuron** [comment]
Hi, this proposal is a pattern we also want to support. Besides collecting and rendering multiple repos to one service to have the deployed state of the service audited at every point in time. Taking this idea even further you could render all manifests to a cluster repo to archive the same on cluster level.

**Comment by mihaigalos** [comment]
I'm tentatively contributing a WIP PoC in https://github.com/akuity/kargo/pull/1447.

**Comment by dan1el-k** [comment]
Hey @krancour,
Is there any update on this ticket ? were you able to discuss this further in the round of maintainers ?
Thanks.

**Comment by krancour** [comment]
In v0.9.0, we plan to break up our "russian nesting doll" of opinionated promotion mechanisms into an ordered sequence of more granular, more explicit, user-defined steps. It is largely a guess, but this issue may be solved incidentally by those plans.

**Comment by krancour** [comment]
Following up on previous comment. I do expect that this will be addressed naturally by #2219

**Comment by krancour** [comment]
The new promotion steps feature can support this.
